research/deep_speech/deep_speech.py

```diff
@@ -41,53 +42,86 @@
 _CER_KEY = "CER"
 
 
-def evaluate_model(
-    estimator, batch_size, speech_labels, targets, input_fn_eval):
+def compute_length_after_conv(args):
+  """Computes the time_steps/ctc_input_length after convolution.
+
+  Suppose that the original feature contains two parts:
+  1) Real spectrogram signals, spanning input_length steps.
+  2) Padded part with all 0s.
+  The total length of those two parts is denoted as max_time_steps, which is
+  the padded length of the current batch. After convolution layers, the time
+  steps of a spectrogram feature will be decreased. As we know the percentage
+  of its original length within the entire length, we can compute the time steps
+  for the signal after conv as follows (using ctc_input_length to denote):
+  ctc_input_length = (input_length / max_time_steps) * output_length_of_conv.
+  This length is then fed into ctc loss function to compute loss.
+
+  Args:
+    args: the arguments to compute ctc_input_length.
+
+  Returns:
+    the ctc_input_length after convolution layer.
+  """
+  max_time_steps, ctc_time_steps, input_length = args
+  ctc_input_length = tf.multiply(
+      tf.to_float(input_length), tf.to_float(ctc_time_steps))
+  ctc_input_length = tf.to_int32(tf.floordiv(
+      ctc_input_length, tf.to_float(max_time_steps)))
+  return ctc_input_length
+
+
+def ctc_loss(args):
+  """Computes the ctc loss for the current batch of predictions."""
+  label_length, ctc_input_length, labels, logits = args
+  label_length = tf.to_int32(tf.squeeze(label_length))
+  ctc_input_length = tf.to_int32(tf.squeeze(ctc_input_length))
+  sparse_labels = tf.to_int32(
+      tf.keras.backend.ctc_label_dense_to_sparse(labels, label_length))
+  y_pred = tf.log(tf.transpose(
+      logits, perm=[1, 0, 2]) + tf.keras.backend.epsilon())
+
+  return tf.expand_dims(
+      tf.nn.ctc_loss(labels=sparse_labels, inputs=y_pred,
+                     sequence_length=ctc_input_length),
+      axis=1)
+
+
+def evaluate_model(estimator, speech_labels, entries, input_fn_eval):
   """Evaluate the model performance using WER anc CER as metrics.
 
   WER: Word Error Rate
   CER: Character Error Rate
 
   Args:
     estimator: estimator to evaluate.
-    batch_size: size of a mini-batch.
     speech_labels: a string specifying all the character in the vocabulary.
-    targets: a list of list of integers for the featurized transcript.
+    entries: a list of data entries (audio_file, file_size, transcript) for the
+      given dataset.
     input_fn_eval: data input function for evaluation.
 
   Returns:
     Evaluation result containing 'wer' and 'cer' as two metrics.
   """
   # Get predictions
-  predictions = estimator.predict(
-      input_fn=input_fn_eval, yield_single_examples=False)
+  predictions = estimator.predict(input_fn=input_fn_eval)
```

Comment:
What's the reason to use predict instead of eval method here? Is that because you can generate proper eval metric here?

Reply:
Yeah, we will use the result (logits) from predict to compute metrics ourselves. Since we haven't implemented this in eval mode, so here we will just call predict.